include/pmacc/math/complex/Bessel.tpp

```diff
@@ -0,0 +1,276 @@
+ *  arguments. Updated Butterworth and Bessel filter tables with files of
+ *   extended parameters including polynomials, poles and component values.
+ *  6/04 -- Revised bessel.zip to correct errors in complex Bessel functions.
+ *
+ *  Further (re-)implementation of this code has been done by FZ Juelich.
+ *  URL: http://apps.jcns.fz-juelich.de/redmine/issues/569#change-2056
+ *  Above URL also includes a accuracy test report of this code against
+ *  SLATEC, MAPLE and MATHEMATICA.
+ */
+
+#pragma once
+
+#include "pmacc/types.hpp"
+#include "pmacc/algorithms/math.hpp"
+#include "pmacc/algorithms/TypeCast.hpp"
+#include "pmacc/math/Complex.hpp"
+
+#include <cmath>
+
+
+namespace pmacc
+{
+namespace algorithms
+{
+namespace math
+{
+namespace bessel
+{
+    namespace pmMath = pmacc::algorithms::math;
+
+    template<
+        typename T_Type,
+        typename T_TableA,
+        typename T_TableB,
+        typename T_TableA1,
+        typename T_TableB1
+    >
+    struct Cbesselj0Base
+    {
+        using result = pmacc::math::Complex< T_Type >;
+        using complex_T = pmacc::math::Complex< T_Type >;
+        using float_T = T_Type;
+        T_Type const eps = float_T( 1.0e-15 );
+        T_Type const el = float_T( 0.5772156649015329 );
+
+        complex_T const cii = complex_T( 0, 1 );
+        complex_T const cone = complex_T( 1, 0 );
+        complex_T const czero = complex_T( 0, 0 );
+
+        HDINLINE result operator( )( complex_T const & z )
+        {
+            T_TableA a;
+            T_TableB b;
+            T_TableA1 a1;
+            T_TableB1 b1;
+            result cj0;
+
+            float_T const a0 = pmMath::abs( z );
+            float_T const z2 = z * z;
+            complex_T z1 = z;
+            if ( a0 == float_T( 0.0 ) )
+            {
+                cj0 = cone;
+                return cj0;
+            }
+            if ( z.get_real() < float_T( 0.0 ) ) z1 = -z;
+            if ( a0 <= float_T( 12.0 ) )
+            {
+                cj0 = cone;
+                complex_T cr = cone;
+                for ( uint32_t k = 1u; k <= 40u; k++ )
+                {
+                    cr *= float_T( -0.25 ) * z2 / float_T( k * k );
+                    cj0 += cr;
+                    if ( pmMath::abs( cr ) < pmMath::abs( cj0 ) * eps ) break;
+                }
+            }
+            else {
+                uint32_t kz;
+                if ( a0 >= float_T( 50.0 ) ) kz = 8u;      // can be changed to 10
+                else if (a0 >= float_T( 35.0 ) ) kz = 10u; //   "      "     "  12
+                else kz = 12u;                             //   "      "     "  14
+                complex_T ct1 = z1 - float_T( M_PI_4 );
+                complex_T cp0 = cone;
+                for ( uint32_t k = 0u; k < kz; k++ )
+                {
+                    cp0 += a[ k ] * pow(
+                        z1,
+                        float_T( -2.0 * k -2.0 )
```

Comment:
please write `float_T( -2.0 ) * float_T( k ) - float_T( 2.0 )` to avoid calculations with the wrong precision